shared/actions/login/saga.js

```diff
@@ -0,0 +1,588 @@
+    }
+  }
+
+  const finishedSaga = function * () {
+    yield call(onBackSaga)
+  }
+
+  const chooseDeviceTypeSaga = function * ({response}) {
+    const deviceTypeMap: {[key: string]: any} = {
+      [Constants.codePageDeviceRoleNewComputer]: Types.CommonDeviceType.desktop,
+      [Constants.codePageDeviceRoleNewPhone]: Types.CommonDeviceType.mobile,
+    }
+    const deviceType = deviceTypeMap[role]
+
+    yield call(setCodePageOtherDeviceRole, role)
+    yield result(response, deviceType)
+  }
+
+  const addDeviceSagas = {
+    ...kex2Sagas(onBackSaga, onBackSaga, finishedSaga),
+    'keybase.1.provisionUi.chooseDeviceType': chooseDeviceTypeSaga,
+  }
+
+  const channelConfig = Saga.singleFixedChannelConfig(Object.keys(addDeviceSagas))
+  const addDeviceChanMap = yield call(addDeviceRpc, channelConfig)
+
+  // Returns an Array<Task>
+  yield _.map(addDeviceSagas, (saga, methodName) => Saga.effectOnChannelMap(c => Saga.safeTakeLatestWithCatch(c, finishedSaga, saga), addDeviceChanMap, methodName))
+}
+
+function * reloginSaga ({payload: {usernameOrEmail, passphrase}}: Constants.Relogin) {
+  const finishedSaga = function * ({error}) {
+    if (error) {
+      const message = 'This device is no longer provisioned.'
+      yield put(Creators.loginDone({message}))
+      yield put(Creators.setLoginFromRevokedDevice(message))
+      yield put(navigateTo([loginTab]))
+    } else {
+      yield call(loginSuccess)
+    }
+  }
+
+  const reloginSagas = {
+    'keybase.1.secretUi.getPassphrase': function * ({response}) {
+      yield result(response, {passphrase: passphrase.stringValue(), storeSecret: true})
+    },
+    'finished': finishedSaga,
+  }
+
+  const channelConfig = Saga.singleFixedChannelConfig(Object.keys(reloginSagas))
+  const chanMap = Types.loginLoginProvisionedDeviceRpcChannelMap(
+    channelConfig,
+    {param: {noPassphrasePrompt: false, username: usernameOrEmail}},
+  )
+
+  yield _.map(reloginSagas, (saga, methodName) => Saga.effectOnChannelMap(c => Saga.safeTakeLatest(c, saga), chanMap, methodName))
+}
+
+function * submitForgotPasswordSaga () {
+  yield put({payload: undefined, type: Constants.actionSetForgotPasswordSubmitting})
+
+  const sagas = {
+    finished: function * ({error}) {
+      if (error) {
+        yield put({
+          error: true,
+          payload: error,
+          type: Constants.actionForgotPasswordDone,
+        })
+      } else {
+        yield put({
+          error: false,
+          payload: undefined,
+          type: Constants.actionForgotPasswordDone,
+        })
+      }
+    },
+  }
+
+  const email = yield select(state => state.login.forgotPasswordEmailAddress)
+  const channelConfig = Saga.singleFixedChannelConfig(Object.keys(sagas))
+  const chanMap = Types.loginRecoverAccountFromEmailAddressRpcChannelMap(channelConfig, {param: {email}})
+  yield _.map(sagas, (saga, methodName) => Saga.effectOnChannelMap(c => Saga.safeTakeLatest(c, saga), chanMap, methodName))
```

Comment:
Seems like you do this operation a lot... would it make sense to move into a helper (which might help name/describe what this is doing better)?